### PR TITLE
Fix stablehlo constant folding bugs through patch

### PR DESCRIPTION
**Pull request by shelkesagar29**
This PR provides a patch to fix two issues in StableHlo
upstream,
- If some op is constant folded and op users are certain
stablehlo ops (which don't allow different type for operand
and result) OR ops with different dialect, `tensor.cast`
should be inserted.
MLIR-TensorRT constant folding pass does this but upstream
stablehlo pass doesn't do this, causing error. Some
logic from MLIR-TRT is patched into upstream.
- Upstream `ReshapeOpCanon` pattern to simplify
`stablehlo.reshape` op failed if input to reshape
is `DenseResourceElementsAttr`. This is fixed.
- Many patterns where replacing op with a new op,
without giving consideration to casting. This is 
fixed.

Solves: https://github.com/NVIDIA/TensorRT-Incubator/issues/378

### CODE REVIEW

**Comment by shelkesagar29** [comment]
@christopherbate helper functions are implemented in header itself to unblock Tripy for now.
As soon as you merge PR to update stablehlo, I will need to update this patch as well.
I will then move definitions  to `Utils.cpp` maybe.

**Comment by christopherbate** [comment]
Can you verify that `ninja -C build check-stablehlo-tests` passes? Can that be added to the CI command as well?

If the StableHLO tests passes then we should be good to merge, although ideally we should be adding additional tests to verify the cases this solves (either in MLIR-TRT tests or in stablehlo via the patch)

**Review by christopherbate — APPROVED** [review]

**Comment by shelkesagar29** [comment]
> Can you verify that `ninja -C build check-stablehlo-tests` passes? Can that be added to the CI command as well?
> 
> If the StableHLO tests passes then we should be good to merge, although ideally we should be adding additional tests to verify the cases this solves (either in MLIR-TRT tests or in stablehlo via the patch)

Yes, stablehlo tests do pass with changes. 
I will add tests for next patch that we will create for newer stablehlo version you will be updating.

**Pull request merged**
